File: Design-Patterns-Using-Ruby/ChainOfResponsibility/ChainOfResponsibilityGUI.rb
# Abstract class representing a handler
class Handler
    attr_accessor :parent, :help_text
    
    def initialize(parent)
        @parent = parent
        @help_text = nil
    end
    
    def show_helper_text
        raise NotImplementedError, 'Subclasses must implement this method'
    end
    
    def set_helper_text(text)
        @help_text = text
    end
end
  
# Class representing a container
class Container < Handler
    def initialize(parent)
        super(parent)
    end
    
    def show_helper_text
        if help_text
            puts "Help :: #{help_text}"
        elsif parent
            puts 'Message passed to next in chain by Container'
            parent.show_helper_text
        end
    end
end

# Class representing a button
class Button < Handler
    attr_accessor :label
    
    def initialize(label, parent)
        super(parent)
        @label = label
    end
    
    def show_helper_text
        if help_text
            puts "Help :: #{help_text}"
        elsif parent
            puts "Message passed to next in chain by Button"
            parent.show_helper_text
        end
    end
end

# Class representing a panel
class Panel < Handler
    def initialize
        super(nil)
    end
    
    def show_helper_text
        if help_text
            puts "Help :: #{help_text}"
        elsif parent
            puts "Message passed to next in chain by Panel"
            parent.show_helper_text
        end
    end
end

# Client code
p = Panel.new
p.set_helper_text('Panel help text.')

b1 = Button.new('Ok', p)
b1.set_helper_text('Ok button help text.')

b2 = Button.new('Cancel', p)

b1.show_helper_text
b2.show_helper_text

=begin 
Help :: Ok button help text.
Message passed to next in chain by Button
Help :: Panel help text.
=end

File: Design-Patterns-Using-JavaScript/ChainOfResponsibility/ChainOfResponsibilityGUI.js
class Handler {
    constructor(parent) {
        this.parent = parent;
        this.helpText = null;
    }

    showHelperText() {
        throw new Error("Abstract method: showHelperText");
    }

    setHelperText(text) {
        this.helpText = text;
    }
}

class Container extends Handler {
    constructor(parent) {
        super(parent);
    }

    showHelperText() {
        if (this.helpText !== null) {
            console.log("Help :: " + this.helpText);
        } else if (this.parent !== null) {
            console.log("Message passed to next in chain by Container");
            this.parent.showHelperText();
        }
    }
}

class Button extends Handler {
    constructor(label, parent) {
        super(parent);
        this.label = label;
    }

    showHelperText() {
        if (this.helpText !== null) {
            console.log("Help :: " + this.helpText);
        } else if (this.parent !== null) {
            console.log("Message passed to next in chain by Button");
            this.parent.showHelperText();
        }
    }
}

class Panel extends Handler {
    constructor() {
        super(null);
    }

    showHelperText() {
        if (this.helpText !== null) {
            console.log("Help :: " + this.helpText);
        } else if (this.parent !== null) {
            console.log("Message passed to next in chain by Panel");
            this.parent.showHelperText();
        }
    }
}

// Client code
const p = new Panel();
p.setHelperText("Panel help text.");

const b1 = new Button("Ok", p);
b1.setHelperText("Ok button help text.");

const b2 = new Button("Cancel", p);

b1.showHelperText();
b2.showHelperText();

/*
Help :: Ok button help text.
Message passed to next in chain by Button
Help :: Panel help text.
*/